packages/mui-base/src/utils/OverridableComponent.ts

```diff
@@ -0,0 +1,24 @@
+import * as React from 'react';
+import { OverridableTypeMap } from '@mui/types';
+
+/**
+ * A component whose root component can be controlled explicitly with a generic type parameter.
+ * Adjusts valid props based on the type of `RootComponent`.
+ *
+ * @template TypeMap The interface the defines the props and a default root element of the component.
+ */
+export type OverridableComponent<TypeMap extends OverridableTypeMap> = {
```

Comment:
```suggestion
export type PolymorphicComponent<TypeMap extends OverridableTypeMap> = {
```

@michaldudak what do you think about changing the name, so that there is no confusion with the two other implementations of it, especially as it is not identical?

Reply:
I don't really have a preference here. `PolymorphicComponent` does sound good, though - it communicates more clearly the intent.

Reply:
Done!